File: packages/python-sdk/src/treedx_sdk/adapters/workspaces.py
from __future__ import annotations

from typing import Any

from .common import json_request, segment
from treedx_sdk.transport import Transport


class WorkspacesAdapter:
    def __init__(self, transport: Transport) -> None:
        self.transport = transport

    def create(self, repo_id: str, body: Any) -> Any:
        return json_request(self.transport, "POST", f"/api/v1/repos/{segment(repo_id)}/workspaces", body)

    def get(self, workspace_id: str) -> Any:
        return json_request(self.transport, "GET", f"/api/v1/workspaces/{segment(workspace_id)}")

    def close(self, workspace_id: str, body: Any | None = None) -> Any:
        return json_request(self.transport, "POST", f"/api/v1/workspaces/{segment(workspace_id)}/close", body)

File: packages/elixir-sdk/lib/treedx/adapters/federation/workspaces.ex
defmodule TreeDxSdk.Workspaces do
  @moduledoc false
  alias TreeDxSdk.Adapters.Common

  def create(client, repo_id, body) do
    Common.json_request(
      client,
      :post,
      "/api/v1/repos/" <> Common.segment(repo_id) <> "/workspaces",
      body,
      %{}
    )
  end

  def get(client, workspace_id) do
    Common.json_request(
      client,
      :get,
      "/api/v1/workspaces/" <> Common.segment(workspace_id),
      nil,
      %{}
    )
  end

  def close(client, workspace_id, body \\ %{}) do
    Common.json_request(
      client,
      :post,
      "/api/v1/workspaces/" <> Common.segment(workspace_id) <> "/close",
      body,
      %{}
    )
  end
end
